Fix lost messages after restart with persistent sessions

dbeinder opened the pull request:
Fixes https://github.com/eclipse/mosquitto/issues/769 / https://github.com/eclipse/mosquitto/issues/384 see repro in https://github.com/eclipse/mosquitto/issues/769#issuecomment-580281178 by @jr-dimedis
This stores the last connected username and listener along with persistent client sessions and attempts to reattach ACLs to the client context, so messages can be retained for clean=false sessions.

With `per_listener_settings ` off, reattaching the ACLs is straightforward, and the previous listener port is ignored. That way even when changing ports, no data is lost.
With `per_listener_settings` on, I don't think there is a 'correct' way to handle a changed port-config - I have adopted the same strategy used for persisting the message source listener, simply store the port and try to find matching listener. Only when changing listener ports in the config and restarting some subscriptions may lose their ACLs until the next subscriber connect. 

There was no unused padding in the client chunk, so I had to bump the DB version.
Do the database changes look good? Are there any unintended side-effects; @ralight suspected some in https://github.com/eclipse/mosquitto/issues/769?

For now this is based on master, since develop doesn't build on windows for me (some sigset_t dependency).
Or should I rebase onto fixes?
I haven't touched the tests yet, since they are based on database binaries.

- [X] Have you signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php), using the same email address as you used in your commits?
- [X] Do each of your commits have a "Signed-off-by" line, with the correct email address? Use "git commit -s" to generate this line for you.
- [X] If you are contributing a bugfix, is your work based off the fixes branch?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [X] Have you successfully run `make test` with your changes locally?

ralight commented:
Thanks for looking into this. I think this would be best off based on `fixes`, despite it introducing a db format change. I don't think there are any other consequences of making this change.

In terms of the tests, duplicating `TEST_v5_client` as `TEST_v6_client`, and add [v6-client.txt](https://github.com/eclipse/mosquitto/files/5035086/v6-client.txt) as `v6-client.test-db` - I think I've got it right.

dbeinder commented:
Because the DB version is stored as `v6` in all write tests, and the write tests share binaries with the read tests, I had to update most v5 binaries. The client chunk is the same for all tests, I added username=`usrname` and listener_port=`1883` to all of them. I renamed most of the v5 tests to v6 and kept the v5 client test. The only really new test is `TEST_v6_client`.

I rebased on `fixes`, this now passes all tests on my machine. @ralight 

Unrelated to this PR: When debugging on the fixes branch, I had to add the calls to `fflush()` on `stdout/stderr` [[here](https://github.com/eclipse/mosquitto/commit/54f3b686dc2a52d952ccf03945c19985440eb49c#diff-f909d0b1a298531f76132943dc55f86aL296)], not sure if that was intended with the changes to file logging. In any case, `setvbuf` with `_IOLBF` is not supported on Windows, that should probably have some platform workaround.

ralight commented:
`_IOLBF`, urgh how tedious :+1: 

Thanks for the test updates, it looks good to me.
